pkg/deployment/resources/pod_finalizers.go

```diff
@@ -0,0 +1,179 @@
+		return maskAny(err)
+	}
+	if len(agencyConns) == 0 {
+		log.Debug().Err(err).Msg("No more remaining agents, we cannot delete this one")
+		return maskAny(fmt.Errorf("No more remaining agents"))
+	}
+	if err := agency.AreAgentsHealthy(ctx, agencyConns); err != nil {
+		log.Debug().Err(err).Msg("Remaining agents are not health")
+		return maskAny(err)
+	}
+	// Remaining agents are health, we can remove this one
+	return nil
+}
+
+// inspectFinalizerPodDrainDBServer checks the finalizer condition for drain-dbserver.
+// It returns nil if the finalizer can be removed.
+func (r *Resources) inspectFinalizerPodDrainDBServer(ctx context.Context, log zerolog.Logger, p *v1.Pod, memberStatus api.MemberStatus, updateMember func(api.MemberStatus) error) error {
+	// Inspect member phase
+	if memberStatus.Phase.IsFailed() {
+		log.Debug().Msg("Pod is already failed, safe to remove drain dbserver finalizer")
+		return nil
+	}
+	// Inspect deployment deletion state
+	apiObject := r.context.GetAPIObject()
+	if apiObject.GetDeletionTimestamp() != nil {
+		log.Debug().Msg("Entire deployment is being deleted, safe to remove drain dbserver finalizer")
+		return nil
+	}
+	// Check PVC
+	pvcs := r.context.GetKubeCli().CoreV1().PersistentVolumeClaims(apiObject.GetNamespace())
+	pvc, err := pvcs.Get(memberStatus.PersistentVolumeClaimName, metav1.GetOptions{})
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to get PVC for member")
+		return maskAny(err)
+	}
+	if !k8sutil.IsPersistentVolumeClaimMarkedForDeletion(pvc) {
+		log.Debug().Msg("PVC is not being deleted, so it is safe to remove drain dbserver finalizer")
+		return nil
+	}
+	log.Debug().Msg("PVC is being deleted, so we will cleanout the dbserver first")
+	// Inspect cleaned out state
+	c, err := r.context.GetDatabaseClient(ctx)
+	if err != nil {
+		log.Debug().Err(err).Msg("Failed to create member client")
+		return maskAny(err)
+	}
+	cluster, err := c.Cluster(ctx)
+	if err != nil {
+		log.Debug().Err(err).Msg("Failed to access cluster")
+		return maskAny(err)
+	}
+	cleanedOut, err := cluster.IsCleanedOut(ctx, memberStatus.ID)
+	if err != nil {
+		return maskAny(err)
+	}
+	if cleanedOut {
+		// Cleanout completed
+		if memberStatus.Conditions.Update(api.ConditionTypeCleanedOut, true, "CleanedOut", "") {
+			if err := updateMember(memberStatus); err != nil {
+				return maskAny(err)
+			}
+		}
+		log.Debug().Msg("Server is cleaned out. Save to remove drain dbserver finalizer")
+		return nil
+	}
+	// Not cleaned out yet, check member status
+	if memberStatus.Conditions.IsTrue(api.ConditionTypeTerminated) {
+		log.Warn().Msg("Member is already terminated before it could be cleaned out. Not good, but removing drain dbserver finalizer because we cannot do anything further")
+		return nil
+	}
+	// Ensure the cleanout is triggered
+	log.Debug().Msg("Server is not yet clean out. Triggering a clean out now")
+	if err := cluster.CleanOutServer(ctx, memberStatus.ID); err != nil {
```

Comment:
I cannot immediately see how often this is called. It seems that as long as the cleanout is ongoing, a new cleanout is ordered every time this routine is called, which happens every time `runPodFinalizers` is run. I found a loop doing this but I could not immediately determine how fast this loop could turn. I do not think it is a real problem to call CleanOutServer again, but I have a bad feeling if many of these calls would be done in rapid succession, whilst it is trying to clean itself out.

Reply:
The interval depends on external events.
So while it is unlikely that these calls are made in rapid succession, it is possible.